File: templates/Enterprise-Template/src/typescript/generator-botbuilder-enterprise/generators/dialog/templates/customDialog/_dialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { ComponentDialog,
        DialogTurnResult,
        WaterfallDialog,
        WaterfallStepContext} from 'botbuilder-dialogs';
import { <%=responsesNameClass%> } from './<%=responsesNameFile%>';

export class <%=dialogNameClass%> extends ComponentDialog {

    // Declare here the type of properties
    private static readonly RESPONDER: <%=responsesNameClass%> = new <%=responsesNameClass%>();

    // Initialize the dialog class properties
    constructor() {
        super(<%=dialogNameClass%>.name);
        this.initialDialogId = <%=dialogNameClass%>.name;
        
        const value: ((sc: WaterfallStepContext<{}>) => Promise<DialogTurnResult<any>>)[] = [
            this.end.bind(this)
        ];
    
   // Add here the waterfall dialog 
        this.addDialog(new WaterfallDialog(this.initialDialogId, value));
    }   

    // Add here end dialog waterfall.
    private async end(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        
        return sc.endDialog(<boolean> sc.result);
    }
}     

File: templates/Skill-Template/DialogTemplate/Skills/Dialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using System.Collections.Specialized;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Solutions.Responses;
using $safeprojectname$.Dialogs.Shared;
using $safeprojectname$.ServiceClients;

namespace $rootnamespace$
{
    public class $safeitemname$ : SkillDialogBase
    {
        public $safeitemname$(
            SkillConfigurationBase services,
            ResponseManager responseManager,
            IStatePropertyAccessor<SkillConversationState> conversationStateAccessor,
            IStatePropertyAccessor<SkillUserState> userStateAccessor,
            IServiceManager serviceManager,
            IBotTelemetryClient telemetryClient)
            : base(nameof($safeitemname$), services, responseManager, conversationStateAccessor, userStateAccessor, serviceManager, telemetryClient)
        {
            var dialog = new WaterfallStep[]
            {
                PromptForName,
                EndDialog
            };

            InitialDialogId = nameof($safeitemname$);
            AddDialog(new WaterfallDialog(nameof($safeitemname$), dialog));
            AddDialog(new TextPrompt($safeitemname$Responses.NamePrompt));
        }

        public async Task<DialogTurnResult> PromptForName(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            return await stepContext.PromptAsync($safeitemname$Responses.NamePrompt, new PromptOptions()
            {
                Prompt = ResponseManager.GetResponse($safeitemname$Responses.NamePrompt),
            });
        }

        public async Task<DialogTurnResult> EndDialog(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            await stepContext.Context.SendActivityAsync(ResponseManager.GetResponse($safeitemname$Responses.HaveNameMessage, new StringDictionary { { "Name", (string)stepContext.Result } }));
            return await stepContext.EndDialogAsync();
        }
    }
}